### ISSUE TITLE
Bad error message - Failed to load Starlark extension

### ISSUE THREAD

**Issue by laurentlb** [bazelbuild/bazel#7784]
Reported by @dhalperi (https://github.com/bazelbuild/bazel/issues/5636#issuecomment-474544205)

> @laurentlb - is there any good debugging advice for how to fix/debug related issues in generated code?
> 
> I'm trying to make https://github.com/pubref/rules_maven work with this flag. Fix 2 in the original comment worked for the bzl files that are there. But I can't figure out how to debug issues in generated files -- I just get warnings about cycles in the graph -- and it seems bazel doesn't materialize the generated files somewhere I can look at them (and mutate them).
> 
> E.g., in https://github.com/batfish/batfish @ master using bazel from yesterday HEAD,
> 
> ```
> $ bazel build //... --incompatible_no_transitive_loads=true        
> INFO: Invocation ID: 0a92899e-0907-4642-877e-ae1644a13cf9
> ERROR: Failed to load Starlark extension '@antlr4_runtime//:rules.bzl'.
> It usually happens when the repository is not defined prior to being used.
> This could either mean you have to add the 'antlr4_runtime' repository with a statement like `http_archive` in your WORKSPACE file (note that transitive dependencies are not added automatically), or the repository 'antlr4_runtime' was defined too late in your WORKSPACE file.
> ERROR: cycles detected during target parsing
> INFO: Elapsed time: 0.084s
> INFO: 0 processes.
> FAILED: Build did NOT complete successfully (0 packages loaded)
>     Fetching @antlr4_runtime; Restarting.
> ```
> but no files are materialized (and the bazel-foo symlinks aren't even created) for me to look at.

The error message is bad, but I don't think it's related to the specific flag.
I'm filing a separate issue, hoping that someone else can look at this.

**Comment by dhalperi** [comment]
Any update here? I'd like to make sure we're able to get all our skylark scripts patched before the flag flips again :).

**Comment by aehlig** [comment]
I can't reproduce the badd message. Please provide more information

```
aehlig@aehlig1:/tmp/work27/batfish$ bazel-exp build //... --incompatible_no_transitive_loads=true 
ERROR: /tmp/work27/batfish/WORKSPACE:19:1: file '@org_pubref_rules_maven//maven:rules.bzl' does not contain symbol 'maven_repositories'
ERROR: /tmp/work27/batfish/WORKSPACE:21:1: name 'maven_repositories' is not defined
ERROR: /tmp/work27/batfish/WORKSPACE:27:1: name 'maven_repository' is not defined
ERROR: Error evaluating WORKSPACE file
ERROR: error loading package '': Encountered error while reading extension file 'rules.bzl': no such package '@antlr4_runtime//': error loading package 'external': Could not load //external package
ERROR: error loading package '': Encountered error while reading extension file 'rules.bzl': no such package '@antlr4_runtime//': error loading package 'external': Could not load //external package
INFO: Elapsed time: 0.076s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded)
aehlig@aehlig1:/tmp/work27/batfish$ ed /tmp/message.txt
```

And that seems to a a reasonable error message related to `--incompatible_no_transitive_loads=true`.

**Comment by aehlig** [comment]
Here `bazel-exp` was a snapshot of bazel at current head.

**Comment by dhalperi** [comment]
@aehlig - the message @laurentlb forked was not from head batfish. As I mentioned in that thread, I was able to fix some of the issues bot not others by following Bazel output. You just saw the first one.

Please try with https://github.com/dhalperi/batfish/tree/bazel-loads:

---

Using bazel built at 75b7ed4a2b07b180454c82c41e2f36a5cbce6a97

and https://github.com/dhalperi/batfish/commit/0fdc3d2ddf3a85679c701f6ce44d9b57ebb97db1 👍 

```
➜  batfish git:(bazel-loads) bazel build //tools/...
INFO: Invocation ID: 0704e24a-bd81-4614-be3d-f972a0b2ec6b
ERROR: Failed to load Starlark extension '@antlr4_runtime//:rules.bzl'.
It usually happens when the repository is not defined prior to being used.
Maybe repository 'antlr4_runtime' was defined later in your WORKSPACE file?
ERROR: cycles detected during target parsing
INFO: Elapsed time: 0.066s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded)
```

**Comment by dhalperi** [comment]
FWIW, the only diff between the two is in an external repostiory (https://github.com/pubref/rules_maven) that I have forked and am trying to get working with Bazel so I can upstream the fixes:

```
commit 5ca3f15c18a694ee789040eef8ca96f414d222f8 (HEAD -> bazel-loads, origin/bazel-loads)
Author: Daniel Halperin <dan@intentionet.com>
Date:   Wed Mar 27 10:37:26 2019 -0700

    maven/rules.bzl: expose rules

diff --git a/maven/rules.bzl b/maven/rules.bzl
index 62c0659..e6fe273 100644
--- a/maven/rules.bzl
+++ b/maven/rules.bzl
@@ -1,3 +1,7 @@
-load("//maven:internal/maven_repositories.bzl", "maven_repositories")
-load("//maven:internal/maven_repository.bzl", "maven_repository")
-load("//maven:internal/require.bzl", "require")
+load("//maven:internal/maven_repositories.bzl", _maven_repositories = "maven_repositories")
+load("//maven:internal/maven_repository.bzl", _maven_repository = "maven_repository")
+load("//maven:internal/require.bzl", _require = "require")
+
+maven_repositories = _maven_repositories
+maven_repository = _maven_repository
+require = _require
```

**Comment by dhalperi** [comment]
Now that Bazel 0.24 is out and this flag will be flipped in the next release, I am feeling a bit more pressure to resolve these issues. Any advice on whether/how I can proceed on my own?

**Comment by laurentlb** [comment]
@aehlig I've checked again at head, the error message is still the same.

> ERROR: Failed to load Starlark extension '@antlr4_runtime//:rules.bzl'.

If possible, show the location (line of the `load`).

> It usually happens when the repository is not defined prior to being used.

Usually? What are the other reasons?

> ERROR: cycles detected during target parsing

This message is confusing. I don't think there's actually a cycle here.

**Comment by dhalperi** [comment]
If helpful, the relevant part of the WORKSPACE looks like [this](https://github.com/batfish/batfish/blob/master/WORKSPACE#L26-L37).

```
# ANTLR 4 Runtime
maven_repository(
    name = "antlr4_runtime",
    transitive_deps = [
        "e27d8ab4f984f9d186f54da984a6ab1cccac755e:org.antlr:antlr4-runtime:4.7.2",
    ],
    deps = [
        "org.antlr:antlr4-runtime:4.7.2",
    ],
)

load("@antlr4_runtime//:rules.bzl", "antlr4_runtime_compile")
```

The load here is defined by the `maven_repository` rule immediately before it. I do not know how to see:

* whether that `maven_repository` rule succeeded or failed
* what that `maven_repository` rule produced, if it succeeded
* what is in the `@antlr4_runtime` space, if anything -- this failure causes the entire `bazel-<foo>` subtrees to not be created (more likely, to be deleted before I have a chance to look).

How do folks debug these types of issues?

**Comment by aehlig** [comment]
> > ERROR: cycles detected during target parsing
> 
> This message is confusing. I don't think there's actually a cycle here.

Note that this message is from Skyframe and technically correct: if a repository is to be loaded, that depends on knowing the definition of the repository; searching for the definition of a repository works by evaluating the `WORKSPACE` file chunk by chunk till the definition is found&mdash;and evaluating that chunk requires evaluating the `load` statement first.

**Comment by dhalperi** [comment]
Thanks! That unblocked me.

Can you please share the CLI flags you enabled to get the output you posted?

**Comment by aehlig** [comment]
> Can you please share the CLI flags you enabled to get the output you posted?

Unfortunately not; that was a quick and dirty `printf` debugging by adding an
```
eventHandler.handle(Event.info("cycle=" + cycle));
eventHandler.handle(Event.info("pathToCycle=" + pathToCycle));
```
into the appropriate [cycle reporter](https://github.com/bazelbuild/bazel/blob/843d18c5292b72d5a7bddc4aa9a9a8947cb57a3c/src/main/java/com/google/devtools/build/lib/skyframe/SkylarkModuleCycleReporter.java#L160); but I'm currently working on a patch to properly report the chain of repositories that lead to the missing one.

**Comment by aehlig** [comment]
Currently under review https://bazel-review.googlesource.com/c/bazel/+/95850

**Comment by aehlig** [comment]
Follow-up change to also record the location of the definition of the repository rule (e.g., to see where the implicit dependencies come from): https://bazel-review.googlesource.com/c/bazel/+/95910

**Comment by nbeloglazov** [comment]
I'm seeing this error and still have no idea how to debug it:
```
ERROR: Failed to load Starlark extension '@rules_nodejs//nodejs:yarn_repositories.bzl'.
Cycle in the workspace file detected. This indicates that a repository is used prior to being defined.
The following chain of repository dependencies lead to the missing definition.
 - @rules_nodejs
This could either mean you have to add the '@rules_nodejs' repository with a statement like `http_archive` in your WORKSPACE file (note that transitive dependencies are not added automatically), or move an existing definition earlier in your WORKSPACE file.
ERROR: cycles detected during target parsing
INFO: Elapsed time: 0.105s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded)
```

I don't know where `@rules_nodejs//nodejs:yarn_repositories.bzl` is coming from. I tried doing `grep  yarn_repositories  -R` from my repo and it didn't return anything. Also `@rules_nodejs` is not used anywhere, instead `build_bazel_rules_nodejs` is used throughout the codebase.

**Comment by poppywu** [comment]
Hi,  @nbeloglazov, I got the same issue as you , did you figure that out?

**Comment by nbeloglazov** [comment]
@poppywu I ended up having to update various nodejs-related rules and packages to be up-to-date.

**Comment by jumbosushi** [comment]
For anyone also confused by this error msg, you can try `bazel sync --only=rules_nodejs` to see where the cycle exists
